File: fsharp/proverb/Proverb.fs
module Proverb

let recite (input: string list): string list =
    let rec reciteRec poem = function
        | [] -> []
        | [_] -> 
            let verse = sprintf "And all for the want of a %s." (Seq.head input) 
            verse::poem
        | head::tail -> 
            let verse = sprintf "For want of a %s the %s was lost." head (Seq.head tail)
            reciteRec (verse::poem) tail

    reciteRec [] input |> List.rev

File: ruby/proverb/proverb.rb
class Proverb
  def initialize(*words, qualifier: nil)
    @words = words
    @qualifier = qualifier
  end

  def to_s
    @words
      .each_cons(2)
      .map { |a, b| "For want of a #{a} the #{b} was lost." }
      .push("And all for the want of a #{@qualifier} #{@words[0]}.".squeeze(' '))
      .join("\n")
  end
end
